packages/gatsby/src/db/index.js

```diff
@@ -0,0 +1,276 @@
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = db.getCollection(type)
+  if (!coll) {
+    coll = db.addCollection(type, { unique: [`id`], indices: [`id`] })
+  }
+
+  return coll.insert(node)
+}
+
+/**
+ * Updates a node in the DB
+ *
+ * @param {Object} node The new node information. Will be merged over
+ * the old one (shallow merge)
+ * @param {Object} oldNode The old node to merge the new node
+ * over. Optional. If not supplied, the old node is found by querying
+ * by node.id
+ */
+function updateNode(node, oldNode) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = db.getCollection(type)
+  if (!coll) {
+    invariant(coll, `${type} collection doesn't exist. When trying to update`)
+  }
+
+  if (!oldNode) {
+    oldNode = getNode(node.id)
+  }
+  const updateNode = _.merge(oldNode, node)
+
+  coll.update(updateNode)
+}
+
+/**
+ * Deletes a node from its type collection.
+ *
+ * @param {Object} the node to delete. Must have an `id`
+ */
+function deleteNode(node) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = db.getCollection(type)
+  if (!coll) {
+    invariant(coll, `${type} collection doesn't exist. When trying to delete`)
+  }
+
+  if (coll.by(`id`, node.id)) {
+    coll.remove(node)
+  } else {
+    console.log(
+      `WARN: deletion of node failed because it wasn't in coll. Node = [${node}]`
+    )
+  }
+}
+
+/////////////////////////////////////////////////////////////////////
+// Queries
+/////////////////////////////////////////////////////////////////////
+
+/**
+ * Returns the node with `id` == id, or null if not found
+ */
+function getNode(id) {
+  invariant(id, `id is null`)
+
+  // We store a collection per node type. So to lookup the node by ID,
+  // we first need to find which collection that node is in, which is
+  // accomplished by iterating through all of them until one is found.
+  //
+  // This is obviously slow. We should consider creating a lookup
+  // collection mapping IDs to their collections. This would require
+  // an additional operation per insert, but would result in faster
+  // lookups
+  const collInfo = _.find(db.listCollections(), collInfo => {
+    const coll = db.getCollection(collInfo.name)
+    return coll.by(`id`, id)
+  })
+  if (collInfo) {
+    const coll = db.getCollection(collInfo.name)
+    return coll.by(`id`, id)
+  } else {
+    return undefined
+  }
+}
+
+/**
+ * Returns all nodes of a type (where typeName == node.internal.type)
+ */
+function getNodesByType(typeName) {
+  const coll = db.getCollection(typeName)
+  if (!coll) return null
```

Comment:
Yeah, maybe just return empty array here? Makes it easier for a consumer of this "API" to handle the empty and null cases the same way.

Reply:
addressed in other comment